src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,253 @@
+            }
+
+            throw lastException;
+        }
+
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, T> funcToExecute, bool forceTryFailedReplicas = false, CancellationToken cancellationToken = default)
+        {
+            var tasks = new List<Task>();
+            var attemptsCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            int replicasTried = 0;
+            Exception lastException = null;
+
+            for (int i = 0; i < _configurationClientAndStates.Count(); i++)
+            {
+                ConfigurationClient client = _configurationClientAndStates.ElementAt(i).Item1;
+                ConfigurationClientState clientState = _configurationClientAndStates.ElementAt(i).Item2;
+
+                if (forceTryFailedReplicas || clientState.IsAvailable())
+                {
+                    var parallelAwait = Task.Delay(this._parallelRetryInterval, attemptsCancellation.Token);
+                    Task<T> funcTask = Task.Run(() => funcToExecute(client), cancellationToken);
+                    tasks.Add(funcTask);
+                    tasks.Add(parallelAwait);
+                    replicasTried++;
+
+                    var completedTask = await Task.WhenAny(tasks);
+                    if (!completedTask.Equals(parallelAwait))
+                    {
+                        if (completedTask.Status == TaskStatus.RanToCompletion)
+                        {
+                            clientState.UpdateConfigurationStoreStatus(requestSuccessful: true);
+
+                            // Safe because the task is completed.
+                            return funcTask.Result;
+                        }
+                        else if (completedTask.Status == TaskStatus.Faulted)
+                        {
+                            tasks.Clear();
+
+                            if (IsRetryableException(completedTask.Exception))
+                            {
+                                lastException = completedTask.Exception;
+                                clientState.UpdateConfigurationStoreStatus(requestSuccessful: false);
+                            }
+                            else
+                            {
+                                throw completedTask.Exception;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            if (replicasTried == 0)
+            {
+                return await ExecuteWithFailOverPolicyAsync<T>(funcToExecute, forceTryFailedReplicas: true, cancellationToken);
+            }
+
+            throw lastException;
+        }
+
+        private bool IsRetryableException(AggregateException ex)
```

Comment:
Is there a way for us to utilize the existing retry policy inside of our configuration clients? They should already contain information for most retriable exceptions.

Reply:
No, we cannot use the existing retry policy. It is not exposed to us.

Reply:
I think the problem is with naming. This is not really about "retryable". This is about "failover-able". This is why Pratik can't find existing reusable logic, because failover is a new concept in this PR. I would suggest to rename the function to "IsFailoverableException".